core/src/main/java/com/linecorp/armeria/server/RoutingDecorator.java

```diff
@@ -0,0 +1,105 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayDeque;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+
+import io.netty.util.AttributeKey;
+
+/**
+ * A pair of a {@link Route} and a {@code decorator} bound to it.
+ */
+final class RoutingDecorator {
+
+    private static final AttributeKey<Queue<Service<HttpRequest, HttpResponse>>> DECORATOR_KEY =
+            AttributeKey.valueOf(RoutingDecorator.class, "ROUTING-DECORATOR");
+
+    private static final DynamicDispatcher dynamicDispatcher = new DynamicDispatcher();
+
+    @Nullable
+    static Function<Service<HttpRequest, HttpResponse>,
+            Service<HttpRequest, HttpResponse>> newDecorator(Router<RoutingDecorator> router) {
+        return delegate -> new InitialDispatcherDecorator(delegate, router);
+    }
+
+    private final Route route;
+    private final Service<HttpRequest, HttpResponse> decorator;
+
+    RoutingDecorator(Route route,
+                     Function<Service<HttpRequest, HttpResponse>,
+                             ? extends Service<HttpRequest, HttpResponse>> decorator) {
+        this.route = requireNonNull(route, "route");
+        this.decorator = requireNonNull(decorator, "decorator").apply(dynamicDispatcher);
+    }
+
+    public Route route() {
+        return route;
+    }
+
+    public Service<HttpRequest, HttpResponse> decorator() {
+        return decorator;
+    }
+
+    static class InitialDispatcherDecorator extends SimpleDecoratingHttpService {
+
+        private final Router<RoutingDecorator> router;
+
+        InitialDispatcherDecorator(Service<HttpRequest, HttpResponse> delegate,
+                                   Router<RoutingDecorator> router) {
+            super(delegate);
+            this.router = router;
+        }
+
+        @Override
+        public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+            final Queue<Service<HttpRequest, HttpResponse>> dynamicDecorators =
+                    router.findAll(ctx.routingContext()).stream()
+                          .filter(Routed::isPresent)
+                          .map(routed -> routed.value().decorator())
+                          .filter(Objects::nonNull)
+                          .collect(Collectors.toCollection(ArrayDeque::new));
+
+            if (dynamicDecorators.isEmpty()) {
+                return delegate().serve(ctx, req);
+            }
+            dynamicDecorators.offer(delegate());
```

Comment:
`.add()` is preferred because we do not worry about a return value unlike `offer()`